kotlinx-coroutines-test/common/src/TestBuilders.kt

```diff
@@ -41,70 +42,250 @@ import kotlin.coroutines.*
+                        // we expect these and will instead throw a more informative exception just below.
+                    }
+                    throw UncompletedCoroutinesError("The test coroutine was not completed after waiting for $dispatchTimeoutMs ms")
                 }
-                scheduler = ctxScheduler
             }
-            dispatcher
         }
-        null -> {
-            scheduler = get(TestCoroutineScheduler) ?: TestCoroutineScheduler()
-            TestCoroutineDispatcher(scheduler)
+        testScope.getCompletionExceptionOrNull()?.let {
+            try {
+                testScope.cleanupTestCoroutines()
+            } catch (e: UncompletedCoroutinesError) {
+                // it's normal that some jobs are not completed if the test body has failed, won't clutter the output
+            } catch (e: Throwable) {
+                it.addSuppressed(e)
+            }
+            throw it
         }
-        else -> throw IllegalArgumentException("Dispatcher must implement TestDispatcher: $dispatcher")
+        testScope.cleanupTestCoroutines()
     }
-    val exceptionHandler = get(CoroutineExceptionHandler).run {
-        this?.let { require(this is UncaughtExceptionCaptor) { "coroutineExceptionHandler must implement UncaughtExceptionCaptor: $this" } }
-        this ?: TestCoroutineExceptionHandler()
-    }
-    val job = get(Job) ?: SupervisorJob()
-    return Pair(this + scheduler + dispatcher + exceptionHandler + job, dispatcher)
+}
+
+/**
+ * Runs [testProcedure], creating a [TestResult].
+ */
+@Suppress("NO_ACTUAL_FOR_EXPECT") // actually suppresses `TestResult`
+internal expect fun createTestResult(testProcedure: suspend () -> Unit): TestResult
+
+/**
+ * Runs a test in a [TestCoroutineScope] based on this one.
+ *
+ * Calls [runTest] using a coroutine context from this [TestCoroutineScope]. The [TestCoroutineScope] used to run
+ * [block] will be different from this one, but will use its [Job] as a parent; therefore, even if calling
+ * [TestCoroutineScope.cleanupTestCoroutines] on this scope were to complete its job, [runTest] won't complete it at the
+ * end of the test.
+ *
+ * Since this function returns [TestResult], in order to work correctly on the JS, its result must be returned
+ * immediately from the test body. See the docs for [TestResult] for details.
+ */
+@ExperimentalCoroutinesApi
+public fun TestCoroutineScope.runTest(
+    dispatchTimeoutMs: Long = DEFAULT_DISPATCH_TIMEOUT_MS,
+    block: suspend TestCoroutineScope.() -> Unit
+): TestResult =
+    runTest(coroutineContext, dispatchTimeoutMs, block)
+
+/**
+ * Run a test using this [TestDispatcher].
+ *
+ * A convenience function that calls [runTest] with the given arguments.
+ *
+ * Since this function returns [TestResult], in order to work correctly on the JS, its result must be returned
+ * immediately from the test body. See the docs for [TestResult] for details.
+ */
+@ExperimentalCoroutinesApi
+public fun TestDispatcher.runTest(
```

Comment:
It has its uses. People also create dispatchers in advance to perform DI.

In any case, this method looks fairly benign and doesn't hurt anything, so a more important question is whether we need `TestCoroutineScope.runTest`, which presupposes that we provide ways to create `TestCoroutineScope`. If we don't, then this convenience method can also go.

Reply:
Let's make this decision in the final branch along with the migration guide.

I still believe `TestCoroutineScope.runTest` may be useful for the examples like this one:
```
@Before 
fun before() {
    myScope = TestCoroutineScope(...)
    myEntity = MyEntity(..., myScope)
}

@Test
fun scenario1() = myScope.runTest {
    myEntity.foo()
    testScheduler.advanceUntilIdle()
    verifyFoo()
}

@Test
fun scenario2() = myScope.runTest {
    myEntity.bar()
    testScheduler.advanceUntilIdle()
    verifyBar()
}
```

_Maybe_ `TestDispatcher.runTest` has the same uses, I was thinking of removing it for the sake of keeping API compact -- it's easily emulated with `runTest(dispatcher)` and only provides one more method to be aware of/to study.

Reply:
I should note that the scope inside `myScope.runTest { ... }` is not the same thing as `myScope`, which may reduce the usefulness of the pattern.

Also, this code is missing a
```kotlin
@AfterTest
fun after() {
  myScope.cleanupTestCoroutines()
}
```
I think this is a major point. According to the comments at the bottom of https://youtrack.jetbrains.com/issue/KT-19813, `AfterTest` and `BeforeTest` don't properly allow returning a `Promise`, which means that either
* the cleanup procedure also can't return a `TestResult`, or
* we shouldn't promote cleaning up in `AfterTest`, but then we'll also have to avoid promoting initializing scopes in `BeforeTest` for symmetry.